lib/gateway.js

```diff
@@ -379,19 +379,33 @@ async function buildGateway (gatewayOpts, app) {
     serviceMap,
     entityResolversFactory: factory,
     pollingInterval: gatewayOpts.pollingInterval,
-    async refresh () {
+    async refresh (isRetry) {
+      const failedMandatoryServices = []
       if (this._serviceSDLs === undefined) {
         this._serviceSDLs = serviceSDLs.join('')
       }
 
       const $refreshResult = await allSettled(
         Object.values(serviceMap).map((service) =>
           service.refresh().catch((err) => {
-            errorHandler(err, service)
+            // If non-mandatory service or if retry count has exceeded for mandatory service then throw
+            if (!service.mandatory || !isRetry) {
+              errorHandler(err, service)
+            }
+
+            // If service is mandatory and retry count has not exceeded then add to service to
+            // failedMandatoryServices so it can be returned for retrying
+            if (service.mandatory) {
+              failedMandatoryServices.push(service)
+            }
           })
         )
       )
 
+      if (failedMandatoryServices.length > 0) {
+        return { schema: null, failedMandatoryServices }
```

Comment:
Can you please avoid this breaking change? Instead of chancing the return type, you can throw an error here.